docs/proposals/scheduling/multi-scheduling-group/README.md

```diff
@@ -0,0 +1,366 @@
+	// group. So that the Karmada scheduler would prefer to schedule workloads
+	// to the main group and the second group will only be considered in case of
+	// the main group does not satisfy restrictions(like, lack of resources).
+	//
+	// Potential use case 2:
+	// For the disaster recovery scenario, the clusters could be organized to
+	// primary and backup groups, the workloads would be scheduled to primary
+	// clusters firstly, and when primary cluster fails(like data center power off),
+	// Karmada scheduler could migrate workloads to the backup clusters.
+	//
+	// +optional
+	OrderedClusterAffinities []OrderedClusterAffinityTerm `json:"orderedClusterAffinities,omitempty"`
+
+	// ClusterTolerations represents the tolerations.
+	// +optional
+	ClusterTolerations []corev1.Toleration `json:"clusterTolerations,omitempty"`
+
+	// SpreadConstraints represents a list of the scheduling constraints.
+	// +optional
+	SpreadConstraints []SpreadConstraint `json:"spreadConstraints,omitempty"`
+
+	// ReplicaScheduling represents the scheduling policy on dealing with the number of replicas
+	// when propagating resources that have replicas in spec (e.g. deployments, statefulsets) to member clusters.
+	// +optional
+	ReplicaScheduling *ReplicaSchedulingStrategy `json:"replicaScheduling,omitempty"`
+}
+// OrderedClusterAffinityTerm selects a set of clusters and indicates the order
+// that the Karmada scheduler should inspect.
+type OrderedClusterAffinityTerm struct {
+	ClusterAffinity `json:",inline"`
+
+	// Order of the cluster group that the Karmada scheduler should inspect with, in the range 1-100.
+	// +required
+	Order int32 `json:"order"`
+	// AffinityName is the name of the cluster group.
+	// +required
+    AffinityName string `json:"affinityName"`
+}
+```
+Each affinity term essentially is the current `ClusterAffinity` but has an
+ordinal number named `order` against which the scheduler evaluates each term individually.
+That means the scheduler always evaluates the first term(with `order: 1`), and turns
+to the next term if the term does not satisfy restrictions.
+
+The following configuration sample declares 3 affinity terms:
+```yaml
+apiVersion: policy.karmada.io/v1alpha1
+kind: PropagationPolicy
+metadata:
+  name: nginx
+spec:
+  resourceSelectors:
+    - apiVersion: apps/v1
+      kind: Deployment
+      name: nginx
+  placement:
+    orderedClusterAffinities:
+      - affinityName: dc-shanghai
+        order: 1
+        clusterNames:
+          - unavailable
+      - affinityName: dc-beijing
+        order: 2
+        clusterNames:
+          - member1
+      - affinityName: dc-hongkong
+        order: 3
+        clusterNames:
+          - member2
+```
+
+During the scheduling phase, the scheduler will first look at the affinity
+term with `order 1` and try to select a feasible cluster set from it.
+
+If no feasible cluster is found from the term, the scheduler will try against the
```

Comment:
It seems like all affinity terms are independent. How could we cope with the case below?

Goal: I would like to deploy 10 replicas among `dc-shanghai` as many as possible, `dc-beijing` is my second choice.
Situation: Only 8 replicas could `dc-shanghai` produce, the other 2 replicas must be deployed among `dc-beijing` which has sufficient resources.

If the affinities are independent, it seems like `dc-shanghai` does not fit so that we could only choose `dc-beijing` to produce all 10 replicas which may not meet the initial requirement.

Reply:
> So I suggest we could make a concept of hierarchical relationship, i.e., if dc-shanghai does not fit, we add dc-beijing with dc-shanghai as an input.

For this case, I think it's not a good idea to group clusters by category `data center`. We can combine two data centers into a new group, E.g.:

```yaml
  placement:
    orderedClusterAffinities:
      - affinityName: primary
        order: 1
        clusterNames:
          - cluster-in-dc-beijing
      - affinityName: backup
        order: 2
        clusterNames:
          - cluster-in-dc-beijing  # still in the candidate list
          - cluster-in-dc-shanghai
```

This proposal does not help to assign more replicas to `cluster-in-dc-beijing`, my idea is we can do that by `weightPreference` or something else. The current `ClusterAffinity` also has this problem.
Please see `Non-Goals` section, that is `The relative priority of clusters in the same group`.